File: test/session.test.ts
import assert from 'node:assert';
import { join } from 'node:path';
import { describe, it } from 'node:test';
import { CHECK_CODE, createSession, FILE_FORMAT, loadTranslations } from '../dist/index.js';

const samplePath = (name: string) => join('test', 'samples', name);

describe('loadTranslations', () => {
	it('scans once and holds what it read', async () => {
		const session = await loadTranslations(samplePath('locales-all-issues'), { target: 'en' });

		assert.deepStrictEqual([...session.locales].sort(), ['en', 'ko']);
		assert.deepStrictEqual(session.groups, ['']);
		assert.strictEqual(session.fileFormat, FILE_FORMAT.SINGLE);
		assert.strictEqual(session.files.length, 2);
		assert.strictEqual(session.get('en', 'greeting'), 'Hello {name}');
	});

	it('flattens on load, so keys are read in their dotted form', async () => {
		const session = await loadTranslations(samplePath('locales-no-issue'), { target: 'en' });

		assert.strictEqual(session.get('ko', 'attr.folder'), '폴더');
		assert.ok(session.keys().includes('desc.hello'));
	});

	it('analyses without reading the files again', async () => {
		const session = await loadTranslations(samplePath('locales-all-issues'), { target: 'en' });
		const first = session.analyze();
		const second = session.analyze();

		assert.strictEqual(first.success, false);
		assert.strictEqual(first.issues.length, second.issues.length);
		assert.strictEqual(first.fileFormat, FILE_FORMAT.SINGLE);
	});

	it('re-checks only the edited key and reports the new state', async () => {
		const session = await loadTranslations(samplePath('locales-all-issues'), { target: 'en' });

		assert.deepStrictEqual(
			session.checkKey('greeting').map((issue) => issue.code),
			[CHECK_CODE.NO_INTERPOLATION_KEY]
		);
		assert.deepStrictEqual(session.set('ko', 'greeting', '{name}님 안녕하세요'), []);
		assert.strictEqual(session.get('ko', 'greeting'), '{name}님 안녕하세요');
	});

	it('carries an edit into the next full analysis', async () => {
		const session = await loadTranslations(samplePath('locales-issue-no-key'), { target: 'en' });

		assert.strictEqual(session.analyze().issuesByCode.NO_KEY?.length, 1);

		session.set('ko', 'attr.folder', '폴더');

		assert.strictEqual(session.analyze().issuesByCode.NO_KEY, undefined);
	});

	it('drops a key from one locale or from all of them', async () => {
		const session = await loadTranslations(samplePath('locales-no-issue'), { target: 'en' });

		assert.deepStrictEqual(
			session.remove('attr.folder', { locale: 'ko' }).map((issue) => issue.code),
			[CHECK_CODE.NO_KEY]
		);

		session.remove('attr.folder');

		assert.strictEqual(session.get('en', 'attr.folder'), undefined);
		assert.deepStrictEqual(session.checkKey('attr.folder'), []);
	});

	it('reload throws away the edits and reads the directory again', async () => {
		const session = await loadTranslations(samplePath('locales-no-issue'), { target: 'en' });

		session.set('ko', 'attr.folder', 'edited');
		assert.strictEqual(session.get('ko', 'attr.folder'), 'edited');

		await session.reload();

		assert.strictEqual(session.get('ko', 'attr.folder'), '폴더');
	});

	it('reports a missing path instead of throwing', async () => {
		const session = await loadTranslations();

		assert.strictEqual(session.path, '');
		assert.strictEqual(session.analyze().success, false);
		assert.strictEqual(session.analyze().issues[0].code, CHECK_CODE.INVALID_OPTIONS);
	});
});

describe('createSession', () => {
	const groups = {
		'common.json': { en: { ok: 'OK' }, ko: { ok: '확인' } },
		'errors.json': { en: { failed: 'Failed' }, ko: {} }
	};

	it('takes translations that are already in memory', () => {
		const session = createSession({ groups }, { target: 'en' });

		assert.deepStrictEqual(session.groups, ['common.json', 'errors.json']);
		assert.strictEqual(session.analyze().issuesByCode.NO_KEY?.length, 1);
	});

	it('finds the group a key lives in, so simple calls need no group', () => {
		const session = createSession({ groups }, { target: 'en' });

		assert.strictEqual(session.get('ko', 'ok'), '확인');
		assert.strictEqual(session.get('en', 'failed'), 'Failed');
		assert.strictEqual(session.checkKey('failed')[0].group, 'errors.json');
	});

	it('writes into the named group', () => {
		const session = createSession({ groups: structuredClone(groups) }, { target: 'en' });

		session.set('ko', 'failed', '실패');

		assert.strictEqual(session.get('ko', 'failed'), '실패');
		assert.strictEqual(session.translations('errors.json').ko.failed, '실패');
		assert.strictEqual(session.analyze().issuesByCode.NO_KEY, undefined);
	});

	it('adds a locale that was not there before', () => {
		const session = createSession({ groups: structuredClone(groups) }, { target: 'en' });

		session.set('ja', 'ok', 'OK', 'common.json');

		assert.ok(session.locales.includes('ja'));
		assert.strictEqual(
			session.checkKey('ok', 'common.json')[0].code,
			CHECK_CODE.NOT_TRANSLATED_VALUE
		);
	});

	it('reset replaces the data but keeps the options', () => {
		const session = createSession({ groups }, { target: 'ko' });

		session.reset({ locales: { ko: { a: '가' }, en: { a: '가' } } });

		assert.deepStrictEqual(session.groups, ['']);
		assert.strictEqual(session.options.target, 'ko');
		assert.strictEqual(session.analyze().issuesByCode.NOT_TRANSLATED_VALUE?.length, 1);
	});
});

File: packages/dart/test/session_test.dart
import 'package:chki18n/chki18n.dart';
import 'package:test/test.dart';

String samplePath(String name) => 'test/samples/$name';

TranslationGroups sampleGroups() => {
  'common.json': {
    'en': {'ok': 'OK'},
    'ko': {'ok': '확인'},
  },
  'errors.json': {
    'en': {'failed': 'Failed'},
    'ko': <String, Object?>{},
  },
};

void main() {
  group('loadTranslations', () {
    test('scans once and holds what it read', () async {
      final session = await loadTranslations(
        path: samplePath('locales-all-issues'),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(session.locales.toList()..sort(), ['en', 'ko']);
      expect(session.groups, ['']);
      expect(session.fileFormat, Chki18nFileFormat.single);
      expect(session.files.length, 2);
      expect(session.get('en', 'greeting'), 'Hello {name}');
    });

    test('flattens on load, so keys are read in their dotted form', () async {
      final session = await loadTranslations(
        path: samplePath('locales-no-issue'),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(session.get('ko', 'attr.folder'), '폴더');
      expect(session.keys(), contains('desc.hello'));
    });

    test('analyses without reading the files again', () async {
      final session = await loadTranslations(
        path: samplePath('locales-all-issues'),
        options: const Chki18nOptions(target: 'en'),
      );
      final first = session.analyze();
      final second = session.analyze();

      expect(first.success, isFalse);
      expect(first.issues.length, second.issues.length);
      expect(first.fileFormat, Chki18nFileFormat.single);
    });

    test('re-checks only the edited key and reports the new state', () async {
      final session = await loadTranslations(
        path: samplePath('locales-all-issues'),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(
        [for (final issue in session.checkKey('greeting')) issue.code],
        [Chki18nCheckCode.noInterpolationKey],
      );
      expect(session.set('ko', 'greeting', '{name}님 안녕하세요'), isEmpty);
      expect(session.get('ko', 'greeting'), '{name}님 안녕하세요');
    });

    test('carries an edit into the next full analysis', () async {
      final session = await loadTranslations(
        path: samplePath('locales-issue-no-key'),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(session.analyze().of(Chki18nCheckCode.noKey).length, 1);

      session.set('ko', 'attr.folder', '폴더');

      expect(session.analyze().issuesByCode[Chki18nCheckCode.noKey], isNull);
    });

    test('drops a key from one locale or from all of them', () async {
      final session = await loadTranslations(
        path: samplePath('locales-no-issue'),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(
        [for (final issue in session.remove('attr.folder', locale: 'ko')) issue.code],
        [Chki18nCheckCode.noKey],
      );

      session.remove('attr.folder');

      expect(session.get('en', 'attr.folder'), isNull);
      expect(session.checkKey('attr.folder'), isEmpty);
    });

    test('reload throws away the edits and reads the directory again', () async {
      final session = await loadTranslations(
        path: samplePath('locales-no-issue'),
        options: const Chki18nOptions(target: 'en'),
      );

      session.set('ko', 'attr.folder', 'edited');
      expect(session.get('ko', 'attr.folder'), 'edited');

      await session.reload();

      expect(session.get('ko', 'attr.folder'), '폴더');
    });

    test('reports a missing path instead of throwing', () async {
      final session = await loadTranslations();

      expect(session.path, '');
      expect(session.analyze().success, isFalse);
      expect(session.analyze().issues.first.code, Chki18nCheckCode.invalidOptions);
    });
  });

  group('createSession', () {
    test('takes translations that are already in memory', () {
      final session = createSession(
        Chki18nInput(groups: sampleGroups()),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(session.groups, ['common.json', 'errors.json']);
      expect(session.analyze().of(Chki18nCheckCode.noKey).length, 1);
    });

    test('finds the group a key lives in, so simple calls need no group', () {
      final session = createSession(
        Chki18nInput(groups: sampleGroups()),
        options: const Chki18nOptions(target: 'en'),
      );

      expect(session.get('ko', 'ok'), '확인');
      expect(session.get('en', 'failed'), 'Failed');
      expect(session.checkKey('failed').first.group, 'errors.json');
    });

    test('writes into the named group', () {
      final session = createSession(
        Chki18nInput(groups: sampleGroups()),
        options: const Chki18nOptions(target: 'en'),
      );

      session.set('ko', 'failed', '실패');

      expect(session.get('ko', 'failed'), '실패');
      expect(session.translations('errors.json')['ko']!['failed'], '실패');
      expect(session.analyze().issuesByCode[Chki18nCheckCode.noKey], isNull);
    });

    test('adds a locale that was not there before', () {
      final session = createSession(
        Chki18nInput(groups: sampleGroups()),
        options: const Chki18nOptions(target: 'en'),
      );

      session.set('ja', 'ok', 'OK', 'common.json');

      expect(session.locales, contains('ja'));
      expect(session.checkKey('ok', 'common.json').first.code, Chki18nCheckCode.notTranslatedValue);
    });

    test('reset replaces the data but keeps the options', () {
      final session = createSession(
        Chki18nInput(groups: sampleGroups()),
        options: const Chki18nOptions(target: 'ko'),
      );

      session.reset(
        const Chki18nInput(
          locales: {
            'ko': {'a': '가'},
            'en': {'a': '가'},
          },
        ),
      );

      expect(session.groups, ['']);
      expect(session.options.target, 'ko');
      expect(session.analyze().of(Chki18nCheckCode.notTranslatedValue).length, 1);
    });
  });
}
